pkg/lorry/engines/grpc/manager.go

```diff
@@ -0,0 +1,177 @@
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package grpc
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/apecloud/kubeblocks/pkg/constant"
+	"github.com/apecloud/kubeblocks/pkg/lorry/dcs"
+	"github.com/apecloud/kubeblocks/pkg/lorry/engines"
+	"github.com/apecloud/kubeblocks/pkg/lorry/plugin"
+	"github.com/apecloud/kubeblocks/pkg/viperx"
+)
+
+type Manager struct {
+	engines.DBManagerBase
+
+	dbClient plugin.DBPluginClient
+}
+
+func NewManager(properties engines.Properties) (engines.DBManager, error) {
+	logger := ctrl.Log.WithName("GRPC")
+	managerBase, err := engines.NewDBManagerBase(logger)
+	if err != nil {
+		return nil, err
+	}
+
+	managerBase.DBStartupReady = false
+
+	host := viperx.GetString(constant.KBEnvPodIP)
+	if h, ok := properties["host"]; ok && h != "" {
+		host = h
+	}
+	port, ok := properties["port"]
+	if !ok || port == "" {
+		return nil, errors.New("grpc port is not set")
+	}
+	dbClient, err := plugin.NewPluginClient(host, port)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create db plugin client")
+	}
+
+	mgr := &Manager{
+		dbClient:      dbClient,
+		DBManagerBase: *managerBase,
+	}
+
+	return mgr, nil
+}
+
+func (mgr *Manager) IsDBStartupReady() bool {
+	if mgr.DBStartupReady {
+		return true
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
+	defer cancel()
+
+	req := &plugin.IsDBReadyRequest{}
+	resp, err := mgr.dbClient.IsDBReady(ctx, req)
+	if err != nil {
+		return false
+	}
+	if resp.Ready {
+		mgr.DBStartupReady = true
+	}
+	return resp.Ready
+}
+
+// JoinCurrentMemberToCluster provides the following dedicated environment variables for the action:
+//
+// - KB_SERVICE_PORT: The port on which the DB service listens.
+// - KB_SERVICE_USER: The username used to access the DB service with sufficient privileges.
+// - KB_SERVICE_PASSWORD: The password of the user used to access the DB service .
+// - KB_PRIMARY_POD_FQDN: The FQDN of the original primary Pod before switchover.
+// - KB_MEMBER_ADDRESSES: The addresses of all members.
+// - KB_NEW_MEMBER_POD_NAME: The name of the new member's Pod.
+// - KB_NEW_MEMBER_POD_IP: The IP of the new member's Pod.
+func (mgr *Manager) JoinCurrentMemberToCluster(ctx context.Context, cluster *dcs.Cluster) error {
+	req := &plugin.JoinMemberRequest{
+		DbInfo:    &plugin.DBInfo{},
+		NewMember: mgr.CurrentMemberName,
+		Members:   cluster.GetMemberAddrs(),
+	}
+
+	if cluster.Leader != nil && cluster.Leader.Name != "" {
+		leaderMember := cluster.GetMemberWithName(cluster.Leader.Name)
+		req.PrimaryFqdn = cluster.GetMemberAddr(*leaderMember)
+	}
+
+	member := cluster.GetMemberWithName(mgr.CurrentMemberName)
+	if member != nil {
+		req.NewMemberIp = member.PodIP
+	}
+
+	_, err := mgr.dbClient.JoinMember(ctx, req)
```

Comment:
The leader name may not be the real leader. And, for many databases, a replica 'join' the cluster by accessing a HA component, bot by accessing leader.  ETCD and Zookeeper joins by accessing the leader, while TiDB joins by accessing PD.

Reply:
Yes, it is important to note that there may be latency in updating and obtaining the correct leader name. here lorry make its best effort to provide the available information, and the action implement should make a double check. If the provided information is not sufficient, the action implementation should consider obtaining additional information through alternative means such as service reference, and variables.